modules/kafka/src/main/java/org/testcontainers/containers/KafkaContainer.java

```diff
@@ -54,40 +56,67 @@ public KafkaContainer withExternalZookeeper(String connectString) {
-        proxy = new SocatContainer()
-                .withNetwork(getNetwork())
-                .withTarget(KAFKA_PORT, networkAlias)
-                .withTarget(ZOOKEEPER_PORT, networkAlias);
+    @NonNull
+    public synchronized Network getNetwork() {
+        if (super.getNetwork() == null) {
+            // Backward compatibility
+            withNetwork(Network.newNetwork());
+        }
+        return super.getNetwork();
+    }
 
-        proxy.start();
-        withEnv("KAFKA_ADVERTISED_LISTENERS", "BROKER://" + networkAlias + ":9092" + "," + getBootstrapServers());
+    @Override
+    protected void doStart() {
+        withCommand("sleep infinity");
 
-        if (externalZookeeperConnect != null) {
-            withEnv("KAFKA_ZOOKEEPER_CONNECT", externalZookeeperConnect);
-        } else {
+        if (externalZookeeperConnect == null) {
             addExposedPort(ZOOKEEPER_PORT);
-            withEnv("KAFKA_ZOOKEEPER_CONNECT", "localhost:2181");
-            withCommand(
-                "sh",
-                "-c",
-                // Use command to create the file to avoid file mounting (useful when you run your tests against a remote Docker daemon)
-                "printf 'clientPort=2181\ndataDir=/var/lib/zookeeper/data\ndataLogDir=/var/lib/zookeeper/log' > /zookeeper.properties" +
-                    " && zookeeper-server-start /zookeeper.properties" +
-                    " & /etc/confluent/docker/run"
-            );
         }
 
         super.doStart();
     }
 
     @Override
-    public void stop() {
-        Stream.<Runnable>of(super::stop, proxy::stop).parallel().forEach(Runnable::run);
+    @SneakyThrows
+    protected void containerIsStarting(InspectContainerResponse containerInfo) {
+        super.containerIsStarting(containerInfo);
+
+        port = getMappedPort(KAFKA_PORT);
+
+        final String zookeeperConnect;
+        if (externalZookeeperConnect != null) {
+            zookeeperConnect = externalZookeeperConnect;
+        } else {
+            zookeeperConnect = "localhost:" + ZOOKEEPER_PORT;
+
+            // Start ZooKeeper
+            ExecCreateCmdResponse execCreateCmdResponse = dockerClient.execCreateCmd(getContainerId())
+                .withCmd("sh", "-c", "" +
+                    "printf 'clientPort=" + ZOOKEEPER_PORT + "\ndataDir=/var/lib/zookeeper/data\ndataLogDir=/var/lib/zookeeper/log' > /zookeeper.properties\n" +
+                    "zookeeper-server-start /zookeeper.properties\n"
+                )
+                .exec();
+
+            dockerClient.execStartCmd(execCreateCmdResponse.getId()).exec(new ExecStartResultCallback()).awaitStarted(10, TimeUnit.SECONDS);
+        }
+
+        String internalIp = containerInfo.getNetworkSettings().getIpAddress();
+
+        ExecCreateCmdResponse execCreateCmdResponse = dockerClient.execCreateCmd(getContainerId())
+            .withCmd("sh", "-c", "" +
+                "export KAFKA_ZOOKEEPER_CONNECT=" + zookeeperConnect + "\n" +
```

Comment:
Why the `\n` instead of using `&&`?
And we could also just use 
```
KAFKA_ZOOKEEPER_CONNECT=... KAFKA_ADVERTISED_LISTENERS=... /etc/confluent/docker/run
```
without the exports, or?

Reply:
Actually, maybe for legibility, treating this as a multi-line script is easier.

Reply:
@kiview actually, I was printing the script during the debugging, and the multiline exports were useful to see them one by one instead of a single long line :) Since it does not change anything (unless it does?), I would keep the multiline version :)

Reply:
Alright, sounds reasonable.